velox/dwio/parquet/reader/TimestampColumnReader.h

```diff
@@ -75,24 +81,30 @@ class TimestampColumnReader : public IntegerColumnReader {
         continue;
       }
 
-      // Convert int128_t to Timestamp by extracting days and nanos.
       const int128_t encoded = reinterpret_cast<int128_t&>(rawValues[i]);
-      const int32_t days = static_cast<int32_t>(encoded >> 64);
-      uint64_t nanos = encoded & ((((1ULL << 63) - 1ULL) << 1) + 1);
-      const auto timestamp = Timestamp::fromDaysAndNanos(days, nanos);
-
-      nanos = timestamp.getNanos();
-      switch (timestampPrecision_) {
-        case TimestampPrecision::kMilliseconds:
-          nanos = nanos / 1'000'000 * 1'000'000;
-          break;
-        case TimestampPrecision::kMicroseconds:
-          nanos = nanos / 1'000 * 1'000;
-          break;
-        case TimestampPrecision::kNanoseconds:
-          break;
+      if (isInt64_) {
```

Comment:
This is assuming that the underlying precision is always NANOS, instead you'll have to look at the underlying precision with which the parquet timestamp is written and then create a timestamp using helpers like Timestamp::{fromMillis, fromMicros, fromNanos}

Reply:
I think I handle the precision casts below. Do I miss sth else?

Reply:
int64 timestamp has an underlying precision unit - https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#timestamp. The supported units are  MILLIS, MICROS or NANOS.
You will have to check the underlying precision and then construct the timestamp using the helper.

An example where you're implementation fails:
If there was an int64 values like 172800000000 written with precision MICROS and the timestampPrecision_ was MILLIS, then in your case 
seconds = 172800000000/1000 = 172800000
nanos = 172800000000 - 172800000 * 1000 = 0

Here, the seconds value is wrong, because the timestamp value was written in microseconds precision, the right value for seconds should have been 172800

Reply:
I get your point, and this is exactly the bug I found when I tried to reuse the Int96 timestamp unit tests for Int64 timestamp, described in #11607.

In other words, I assume the reader aligns the timestamp precision as the writer.